Feat: migrate to uv from pip

fujikosu opened the pull request:
# Purpose

This PR switches the package management system from pip to uv. uv is a new package manager that is way faster than pip.
uv is built in Rust, enabling it to outperform pip significantly. Benchmarks indicate that uv runs way faster than pip (https://github.com/astral-sh/uv/blob/main/BENCHMARKS.md). This speed is achieved through parallel downloads, efficient caching mechanisms, and by downloading only necessary metadata instead of entire package files.
Unlike pip, uv uses a lockfile (uv.lock) alongside pyproject.toml to ensure consistent and reproducible installations across different environments. This approach minimizes dependency conflicts and enhances reliability, similar to the functionality provided by tools like Poetry.

To support this switch, the following code changes were made:
- CI (GitHub Actions migrated to `astral-sh/setup-uv` and `uvx`; Azure DevOps pipelines updated)
- Dockerfiles (all 3 rewritten to use `uv sync --locked` with bind mounts)
- `requirements.txt` → `pyproject.toml` + `uv.lock` in each `.devcontainer`
- `requirements-dev.txt` removed — dev dependencies are now included in each individual `pyproject.toml`
- `src/common/` removed — shared dependencies are now managed per-project in `pyproject.toml`. The common module pattern was rarely needed in practice across multiple projects, and uv works best with all dependencies in a single `pyproject.toml` per environment
- Dependabot (`pip` ecosystem replaced with `uv`)
- `README.md` updated with uv instructions, new "How to update python packages" section, CI badge, and updated directory tree
- `src/sample_pytorch_gpu_project/README.md` updated (`requirements.txt` → `pyproject.toml`)
- PR templates (both GitHub and Azure DevOps) updated to reference `pyproject.toml`
- `devcontainer.json` updated: build context narrowed from repo root to each `.devcontainer/` directory, `remoteEnv` with `UV_PROJECT` added
- `create-env.yaml` comment clarified for build context
- `.gitignore` updated (added `.DS_Store`)
- `.dockerignore` removed (no longer needed since build context changed from repo root to each `.devcontainer/` directory)
- Deprecated/unmaintained VS Code extensions removed from all devcontainer.json files (`github.copilot-chat`, `donjayamanne.python-environment-manager`, `donjayamanne.githistory`, `njpwerner.autodocstring`, `donjayamanne.vscode-default-python-kernel`, `marchiore.csvtomarkdown`)
- Added `.github/copilot-instructions.md` for repo-wide Copilot custom instructions (Google-style docstrings, uv for dependency management)


## Does this introduce a breaking change?

* [x] Yes
* [ ] No

## Author pre-publish checklist

* [x] No PII in logs or output
* [x] Made corresponding changes to the documentation
* [x] All new packages used are included in pyproject.toml
* [x] Functions use type hints, and there are no type hint errors

## Pull Request Type

What kind of change does this Pull Request introduce?

* [ ] Bugfix
* [x] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [x] Documentation content changes
* [ ] Experiment notebook

fujikosu committed:
update tool.ruff target-version to py311

fujikosu committed:
migrate notebook container to uv

fujikosu committed:
migrate sample_cpu_project to uv

fujikosu committed:
fix comment in Dockerfile

fujikosu committed:
remove unused requirements.txt

fujikosu committed:
migrate gpu container to uv

fujikosu committed:
use uv in CI

fujikosu committed:
add cuda version check in sample. switch base container from cuda to cuda-dl-base as required pytorch dependencies like libcusparselt0 was missing

fujikosu committed:
update dockerfiles to use UID 1000

fujikosu committed:
add uv support to dependabot

fujikosu committed:
add uv related docs

fujikosu committed:
add .DS_Store to gitignore

fujikosu committed:
update docs to uv version

fujikosu committed:
tested uv install commands and updated readme

fujikosu committed:
Merge branch 'main' into feat/migrate-to-uv-from-pip

fujikosu committed:
update uv lock procedure doc

fujikosu committed:
update azdo pipelines

fujikosu committed:
fix permission issue for certifi

fujikosu committed:
permission fix

fujikosu committed:
get rid of cuda 12.6 setting to default back to 12.4

fujikosu committed:
updated lockfile

fujikosu committed:
update README to include more mentions about uv and PR template

meciwo reviewed (commented):
Reviewed all files. It looks good to me except `src/common/requirements.txt`

As you mentioned 
># This part can be potentially improved by https://docs.astral.sh/uv/concepts/projects/workspaces/#when-not-to-use-workspaces to move away from requirements.txt and gets its own lock file

What about trying to use workspaces?

fujikosu committed:
update Dockerfiles to change UV_LINK_MODE and make use of mounted cache

fujikosu committed:
fix the mount path in gpu container

fujikosu committed:
Merge branch 'main' into feat/migrate-to-uv-from-pip

fujikosu committed:
build(docker): update Dockerfiles to allow non-root user package management and simplify dependency installation

fujikosu committed:
Update README and environment configuration for Azure ML

- Simplified instructions in README regarding environment updates by removing redundant information about dependencies in requirements.txt.
- Adjusted the build context in create-env.yaml to use the repository root for Dockerfile compatibility, ensuring correct path resolution for dependencies.

fujikosu committed:
build(deps): remove pip updates from Dependabot configuration

fujikosu commented:
> Reviewed all files. It looks good to me except `src/common/requirements.txt`
> 
> As you mentioned
> 
> > # This part can be potentially improved by https://docs.astral.sh/uv/concepts/projects/workspaces/#when-not-to-use-workspaces to move away from requirements.txt and gets its own lock file
> 
> What about trying to use workspaces?

Thanks for your comment! After all, I decided to remove this common module pattern. The common module pattern was rarely needed in practice across multiple projects, and uv works best with all dependencies in a single pyproject.toml per environment

fujikosu committed:
build(pipeline): update README for correct uv command for removing mlflow dep. Get rid of venv creation for azdo linter

fujikosu committed:
build(docs): update README and devcontainer configurations for uv migration

fujikosu committed:
build(docs): update README and Docker configurations for UV integration

fujikosu committed:
build(docker): remove .dockerignore file for UV migration

fujikosu committed:
build(docs): add Copilot instructions and update Python conventions for UV migration

junkataoka commented on notebooks/.devcontainer/Dockerfile:
```diff
@@ -1,4 +1,7 @@
 FROM python:3.11.15
+COPY --from=ghcr.io/astral-sh/uv:latest /uv /uvx /bin/
```
All the Dockerfiles  use `:latest`  tag on uv, but should we pin a tag like `uv:0.5.x` or a digest and let dependabot bumps it?

fujikosu replied:
Thanks! yeah that would be ideal to go with this approach instead. But unfortunately, dependabot still doesn't support parsing of COPY statement. The issue is tracked here
https://github.com/dependabot/dependabot-core/issues/5103

That's the biggest blocker and with these reasonings below, I'd lean toward keeping :latest for now

- uv's CLI surface we rely on (uv sync --locked, the /uv / /uvx binaries) has been stable, and Astral publishes the uv image as a distroless artifact specifically intended to be consumed via COPY --from, so this usage pattern is officially supported.
- Reproducibility of the Python env comes from uv.lock, not the uv binary version. A newer uv against the same lockfile produces the same environment.
- These are dev containers, not deployed artifacts, so the blast radius of a surprise uv bump is small.

junkataoka reviewed (approved):
LGTM. Just one minor comment!
